client/comms/wsconn.go

```diff
@@ -424,28 +420,29 @@ func (conn *wsConn) Stop() {
 // to be down, the message failed to marshall to JSON, or writing to the
 // websocket link failed.
 func (conn *wsConn) Send(msg *msgjson.Message) error {
-	if !conn.isConnected() {
+	if conn.IsDown() {
 		return fmt.Errorf("cannot send on a broken connection")
 	}
 
-	conn.wsMtx.Lock()
-	defer conn.wsMtx.Unlock()
-	err := conn.ws.SetWriteDeadline(time.Now().Add(writeWait))
-	if err != nil {
-		log.Errorf("failed to set write deadline: %v", err)
-		return err
-	}
-
 	// Marshal the Message first so that we don't send junk to the peer even if
 	// it fails to marshal completely, which gorilla/websocket.WriteJSON does.
 	b, err := json.Marshal(msg)
 	if err != nil {
 		log.Errorf("Failed to marshal message: %v", err)
 		return err
 	}
+
+	conn.wsMtx.Lock()
+	defer conn.wsMtx.Unlock()
+	err = conn.ws.SetWriteDeadline(time.Now().Add(writeWait))
+	if err != nil {
+		log.Errorf("Send: failed to set write deadline: %v", err)
+		return err
+	}
+
 	err = conn.ws.WriteMessage(websocket.TextMessage, b)
 	if err != nil {
-		log.Errorf("write error: %v", err)
+		log.Errorf("Send: WriteMessage error: %v", err)
```

Comment:
I had briefly added a `conn.close()` here to cause the read loop to error and trigger reconnect, but the write error itself seems to be enough to cause reads (e.g. `ws.ReadJSON`) to return with an error immediately.